### ISSUE TITLE
ESM3 fails to generate target protein domains sequences

### ISSUE THREAD

**Issue by juan-rodriguez-rivas** [evolutionaryscale/esm#302]
## Description
We are working in a research paper and have observed a significant limitation in ESM3's ability to generate sequences that adhere to provided functional domain annotations. 

Using a test set of 20 distinct protein domains, we provided the model with fully masked sequences paired with the corresponding functional annotations for the target domain. The model successfully generated sequences matching the target domain in only 2 of the 20 cases. In the remaining 18 cases, the generated sequences failed to show any hits for the target domain.

#### 1. Run the provided script
The provided script below generates two FASTA files: one for a "working" case and one for a "failing" case. It uses the `esm3-open` model.

#### 2. Verify results
Submit the generated FASTA files to the [InterPro Sequence Search](https://www.ebi.ac.uk/interpro/search/sequence/) tool to verify domain hits. Download TSV file and check pfam hits.

**Observed Results:**
* **Case interpro IPR001789 (pfam PF00072):** 9/10 sequences returned valid hits (**Success**).
* **Case interpro IPR000504 (pfam PF00076):** 0/10 sequences returned valid hits (**Failure**).

## Minimal example
The following script runs on CPU. Packages required: torch, esm, huggingface_hub

```python
#!/usr/bin/env python
import os
import torch
from huggingface_hub import login
from esm.sdk.api import ESM3InferenceClient, ESMProtein, GenerationConfig
from esm.models.esm3 import ESM3
from esm.utils.types import FunctionAnnotation
import esm.utils.function.interpro

def generate_seqs_domain(model, interpro_id, interpro_name, seq_lengths, output_file):
    generated_seqs = []
    
    for i, seq_length in enumerate(seq_lengths):
        print(f"Generating {interpro_id} {i+1}/{len(seq_lengths)} (len {seq_length})...")
        
        # Prepare prompt and annotations
        prompt = "_" * (seq_length + 1) 
        function_annotation_str = f"{interpro_name} ({interpro_id})"
        annotations = [FunctionAnnotation(label=function_annotation_str, start=2, end=seq_length)]
        
        protein = ESMProtein(sequence=prompt, function_annotations=annotations)
        protein_cod = model.encode(protein)
        
        # Generate sequence
        protein_cod = model.generate(
            protein_cod, 
            GenerationConfig(track="sequence", strategy="random", num_steps=10, temperature=1.0)
        )
        
        protein_dec = model.decode(protein_cod)
        generated_seqs.append(protein_dec.sequence[1:])

    with open(output_file, "w") as f:
        for idx, seq in enumerate(generated_seqs):
            f.write(f">seq{idx+1}_len{len(seq)}\n{seq}\n")

def main():
    torch.manual_seed(1)
    device = torch.device("cpu")
    
    token = os.environ.get("HF_TOKEN")
    if not token: 
        print("Please set HF_TOKEN environment variable.")
        return
    login(token=token)

    print("Loading ESM3 model...")
    model: ESM3InferenceClient = ESM3.from_pretrained("esm3-open", device=device)
    interpro_table = esm.utils.function.interpro.InterPro()

    # Lengths are obtained from the empiral distribution of natural sequences
    test_cases = [
        {"name": "Working Case", "id": "IPR001789", "lengths":[109, 106, 111, 106, 112, 107, 112, 110, 115, 115]}, 


        {"name": "Failing Case", "id": "IPR000504", "lengths":[62, 61, 70, 62, 69, 72, 71, 69, 64, 53]}
    ]

    for case in test_cases:
        entry = interpro_table.entries[case['id']]
        out = f"generated_{case['id']}.fasta"
        generate_seqs_domain(model, case['id'], entry.name, case['lengths'], out)
        print(f"Results saved to {out}")

if __name__ == "__main__":
    main()
```

## Observations

Some test we have performed:
 
-  Versions: The issue persists across all ESM3 versions from v3.0.8 through v3.2.3.
- Prompts: We performed exhaustive trials of various prompting strategies, including varying sequence lengths, adjusting functional annotation delimiters, tuning temperatures, altering sampling strategies, etc.
- Functional annotations: To rule out issues with annotation formatting, we used ESM3 to predict function tracks from natural sequences (it worked perfectly) and used these as prompts for fully masked sequence reconstruction.

In all scenarios, the issue persisted.

The cases where it works are tipically highly populated families such as PF00072. 

## Question

I am inclined to see this as a model limitation regarding functional track adherence to target protein domains, but it could also stem from a bug or a non-obvious prompting requirement. Any insights would be extremely valued. In particular, are there specific recommendations for function-conditioned generation?

**Comment by santiag0m** [comment]
In the ESM3 paper we report results across all models. From the paper you can see the smaller models require a larger generation budget to solve the task, but they do have this capability. 

One can see the task can be accomplished with a smaller generation budget using a guidance technique (see the [guided generation notebook](https://github.com/evolutionaryscale/esm/blob/main/cookbook/tutorials/5_guided_generation.ipynb)).

Below is a script to do generation for the case you mentioned, adapting the function track of the model to do guidance. With it I got the following protein sequence which shows pTM=0.88 and matches for RNA Recognition Motif in InterPro Scan: `QAPPHPNLAFFVGNLPYTTRAEDFRRTLNEHLLVDDPWTLRQIIIYKDIHAFSKAHSCAFVEFDCVKDGHAFMEAVTLKGFRVDKNCYTVSWCDKTKFND`

Please feel free to use this approach if useful!

```python
"""
Tutorial: Function-Guided Generation with ESM3
=================================================
Issue #302: plain function-conditioned generation for some InterPro families
(e.g. IPR000504 / PF00076, the RRM RNA-binding domain) succeeds only ~10% of
the time.  This script fixes that with the same guided-generation infrastructure
from Tutorial 5, using ESM3 itself as the scoring function.

Key insight: ESM3's forward pass returns function_logits even when the function
track is not provided as input.  These logits encode, for each sequence position
and each of the 8 LSH token depths, how strongly the model predicts a given
function token.  By comparing them against the LSH encoding of the target
InterPro domain, we get a non-circular, sequence-derived score in log-probability
space that guided decoding can optimise.

Run:
    python function_guided_generation.py
"""

import torch
import matplotlib.pyplot as plt

from esm.models.esm3 import ESM3
from esm.sdk.api import (
    ESMProtein,
    ESMProteinError,
    FunctionAnnotation,
    GenerationConfig,
)
from esm.sdk.experimental import ESM3GuidedDecoding, GuidedDecodingScoringFunction
from esm.sdk.experimental import (
    ConstraintType,
    ESM3GuidedDecodingWithConstraints,
    GenerationConstraint,
)

# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------
TARGET_INTERPRO = "IPR000504"   # RRM domain (PF00076) – known low-success case
PROTEIN_LENGTH  = 100
N_SAMPLES       = 10
NUM_STEPS       = PROTEIN_LENGTH // 8
NUM_SAMPLES_PER_STEP = 8


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------
def _get_target_function_tokens(model: ESM3, interpro_id: str) -> torch.Tensor:
    """Return the [8] LSH function-token vector for a given InterPro annotation.

    All positions annotated with the same domain share the same token vector, so
    a single-residue reference protein is sufficient.
    """
    ref = ESMProtein(
        sequence="A",
        function_annotations=[FunctionAnnotation(label=interpro_id, start=1, end=1)],
    )
    ref_pt = model.encode(ref)
    assert not isinstance(ref_pt, ESMProteinError)
    assert ref_pt.function is not None
    return ref_pt.function[1]  # [8]  — skip BOS token


def _ipr_log_prob(sequence: str, model: ESM3, target_tokens: torch.Tensor) -> float:
    """Mean log-probability that the model assigns to the target function tokens.

    Calls model.forward() with sequence tokens only (function track ignored /
    defaulting to PAD).  Reads function_logits [L+2, 8, 260] and computes
    the mean log-softmax probability of the target token at each position/depth.

    Score is in (-inf, 0]; higher means the sequence looks more like the target
    domain to the model's function head.
    """
    pt = model.encode(ESMProtein(sequence=sequence))
    assert not isinstance(pt, ESMProteinError)
    assert pt.sequence is not None

    seq_tokens = pt.sequence.unsqueeze(0).to(pt.device)

    with torch.no_grad(), torch.autocast(
        enabled=True, device_type=seq_tokens.device.type, dtype=torch.bfloat16
    ):
        output = model.forward(sequence_tokens=seq_tokens)

    # function_logits: [1, L+2, 8, 260] → strip batch and BOS/EOS → [L, 8, 260]
    logits = output.function_logits[0, 1:-1].float()
    log_probs = torch.log_softmax(logits, dim=-1)  # [L, 8, 260]

    depth_idx = torch.arange(8, device=log_probs.device)
    target = target_tokens.to(log_probs.device)
    # Gather log p(target_token[d]) at each position and depth
    token_log_probs = log_probs[:, depth_idx, target]  # [L, 8]
    return token_log_probs.mean().item()


# ---------------------------------------------------------------------------
# Scoring functions
# ---------------------------------------------------------------------------
class InterProScoringFunction(GuidedDecodingScoringFunction):
    """Score a protein by how strongly ESM3's function head predicts a target domain.

    Uses model.forward() with sequence-only input (function track defaulting to
    PAD — no conditioning).  The score is the mean log-probability of the
    target InterPro domain's LSH token vector, averaged over positions and
    the 8 token depths.  Range: (-inf, 0], higher = more domain-like.
    """

    def __init__(self, model: ESM3, interpro_id: str):
        self.model = model
        self.target_tokens = _get_target_function_tokens(model, interpro_id)

    def __call__(self, protein: ESMProtein) -> float:
        assert protein.sequence is not None
        return _ipr_log_prob(protein.sequence, self.model, self.target_tokens)


class PTMScoringFunction(GuidedDecodingScoringFunction):
    def __call__(self, protein: ESMProtein) -> float:
        assert protein.ptm is not None
        return float(protein.ptm)


def make_protein(length: int, interpro_id: str) -> ESMProtein:
    return ESMProtein(
        sequence="_" * length,
        function_annotations=[FunctionAnnotation(label=interpro_id, start=1, end=length)],
    )


def check_success(
    protein: ESMProtein,
    model: ESM3,
    target_tokens: torch.Tensor,
    threshold: float = -6.3,
) -> bool:
    """Return True if the model's function head predicts the target domain.

    Uses the same _ipr_log_prob score as the scoring function; a threshold of
    -6.3 sits between a genuine RRM sequence (~-5.7) and a random sequence (~-6.6).
    """
    assert protein.sequence is not None
    return _ipr_log_prob(protein.sequence, model, target_tokens) > threshold


def run_batch(generator, n: int, verbose_label: str) -> list[ESMProtein]:
    proteins = []
    for i in range(n):
        p = make_protein(PROTEIN_LENGTH, TARGET_INTERPRO)
        if isinstance(generator, (ESM3GuidedDecoding, ESM3GuidedDecodingWithConstraints)):
            out = generator.guided_generate(
                protein=p,
                num_decoding_steps=NUM_STEPS,
                num_samples_per_step=NUM_SAMPLES_PER_STEP,
                verbose=False,
            )
        else:
            out = generator.generate(
                input=p,
                config=GenerationConfig(track="sequence", num_steps=NUM_STEPS),
            )
        proteins.append(out)
        print(f"  [{verbose_label}] {i+1}/{n}")
    return proteins


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("Loading ESM3-open...")
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model = ESM3.from_pretrained("esm3-open").to(device)

    target_tokens = _get_target_function_tokens(model, TARGET_INTERPRO)
    print(f"IPR000504 LSH tokens: {target_tokens.tolist()}")

    interpro_scorer = InterProScoringFunction(model, TARGET_INTERPRO)

    # 1. Baseline: plain function-conditioned generation
    print("\n--- Baseline ---")
    baseline = run_batch(model, N_SAMPLES, "baseline")

    # 2. Guided: maximise InterPro log-probability
    print("\n--- Guided (InterPro score) ---")
    guided = ESM3GuidedDecoding(client=model, scoring_function=interpro_scorer)
    guided_proteins = run_batch(guided, N_SAMPLES, "guided")

    # 3. Guided + pTM >= 0.7 constraint
    print("\n--- Guided + pTM constraint ---")
    constrained = ESM3GuidedDecodingWithConstraints(
        client=model,
        scoring_function=interpro_scorer,
        constraints=[GenerationConstraint(PTMScoringFunction(), 0.7, ConstraintType.GREATER_EQUAL)],
        damping=1.0,
        learning_rate=10.0,
    )
    constrained_proteins = run_batch(constrained, N_SAMPLES, "constrained")

    # ---------------------------------------------------------------------------
    # Evaluate
    # ---------------------------------------------------------------------------
    strategies = [
        ("Baseline",          baseline),
        ("Guided",            guided_proteins),
        ("Guided + pTM>=0.7", constrained_proteins),
    ]

    successes  = {}
    ipr_scores = {}
    for label, proteins in strategies:
        ipr_scores[label] = [_ipr_log_prob(p.sequence, model, target_tokens) for p in proteins]  # type: ignore[arg-type]
        successes[label]  = [s > -6.3 for s in ipr_scores[label]]

    print(f"\n{'Strategy':<24} {'Success':>10}  {'Mean log-prob':>14}")
    print("-" * 52)
    for label, _ in strategies:
        n_ok  = sum(successes[label])
        score = sum(ipr_scores[label]) / len(ipr_scores[label])
        print(f"{label:<24} {n_ok:>4}/{N_SAMPLES:<5}  {score:.4f}")

    # ---------------------------------------------------------------------------
    # Plot
    # ---------------------------------------------------------------------------
    _, axes = plt.subplots(1, 2, figsize=(11, 4))

    # Left: success rate bar chart
    ax = axes[0]
    labels      = [l for l, _ in strategies]
    n_successes = [sum(successes[l]) for l in labels]
    bars = ax.bar(labels, [n / N_SAMPLES * 100 for n in n_successes],
                  color=["#d62728", "#1f77b4", "#2ca02c"], width=0.5)
    ax.set_ylabel("Success rate (%)")
    ax.set_title(f"Fraction of runs where {TARGET_INTERPRO} is predicted\n(N={N_SAMPLES} per strategy)")
    ax.set_ylim(0, 110)
    for bar, n in zip(bars, n_successes):
        ax.text(bar.get_x() + bar.get_width() / 2, bar.get_height() + 2,
                f"{n}/{N_SAMPLES}", ha="center", va="bottom", fontsize=11)

    # Right: distribution of InterPro log-prob scores
    ax = axes[1]
    colors = ["#d62728", "#1f77b4", "#2ca02c"]
    for (label, _), color in zip(strategies, colors):
        ax.hist(ipr_scores[label], bins=10, alpha=0.6,
                label=label, color=color, edgecolor="white")
    ax.axvline(-6.3, linestyle="--", color="k", label="success threshold")
    ax.set_xlabel(f"Mean log p(IPR000504 tokens | seq)")
    ax.set_ylabel("Count")
    ax.set_title("Distribution of InterPro scores")
    ax.legend()

    plt.tight_layout()
    out_path = "function_guided_generation_results.png"
    plt.savefig(out_path, dpi=150)
    print(f"\nPlot saved to {out_path}")
    plt.show()


if __name__ == "__main__":
    main()
```

**Comment by juan-rodriguez-rivas** [comment]
Thank you very much! The guidance you proposed really helps. The quality (e.g. HMMER e-values/bitscores) of the sequences is still limited on average, but it is much better than before. Much appreciated!
